lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -3047,53 +3050,75 @@ ConformanceChecker::inferTypeWitnessesViaValueWitnesses(ValueDecl *req) {
-                             return false;
-                           if (!dmt->getBase()->isEqual(Conformance->getType()))
-                             return false;
-                           
-                           return true;
-                         });
-                       
-                       if (containsTautologicalType) {
-                         return true;
-                       }
-
-                       // Check that the type witness meets the
-                       // requirements on the associated type.
-                       if (auto failed = checkTypeWitness(TC, DC, result.first,
-                                                          result.second)) {
-                         witnessResult.NonViable.push_back(
-                           std::make_tuple(result.first,result.second,failed));
-                         return true;
-                       }
-
-                       return false;
-                     }),
-      witnessResult.Inferred.end());
+    for (unsigned i = 0; i < witnessResult.Inferred.size(); /*nothing*/) {
+#define REJECT {\
+  witnessResult.Inferred.erase(witnessResult.Inferred.begin() + i); \
+  continue; \
+}
+      auto &result = witnessResult.Inferred[i];
+
+      // Filter out errors.
+      if (result.second->hasError())
+        REJECT;
+
+      // Filter out duplicates.
+      if (!known.insert({result.first, result.second->getCanonicalType()})
+                .second)
+        REJECT;
+     
+      // Filter out circular possibilities, e.g. that
+      // AssocType == S.AssocType or
+      // AssocType == Foo<S.AssocType>.
+      bool containsTautologicalType =
+        result.second.findIf([&](Type t) -> bool {
+          auto dmt = t->getAs<DependentMemberType>();
+          if (!dmt)
+            return false;
+          if (!associatedTypesAreSameEquivalenceClass(dmt->getAssocType(),
+                                                      result.first))
+            return false;
+          if (!dmt->getBase()->isEqual(Conformance->getType()))
+            return false;
+          
+          return true;
+        });
+      
+      if (containsTautologicalType)
+        REJECT;
+      
+      // Check that the type witness doesn't contradict an
+      // explicitly-given type witness. If it does contradict, throw out the
+      // witness completely.
+      if (!allUnresolved.count(result.first)) {
+        auto existingWitness =
+        Conformance->getTypeWitness(result.first, nullptr)
+        .getReplacement();
+        // If the deduced type contains an irreducible
+        // DependentMemberType, that indicates a dependency
+        // on another associated type we haven't deduced,
+        // so we can't tell whether there's a contradiction
+        // yet.
+        auto newWitness = result.second->getCanonicalType();
+        if (!newWitness.findIf([&](Type t) -> bool {
+              return t->is<DependentMemberType>();
```

Comment:
Why not use `hasTypeParameter()` instead of looking for a `DependentMemberType`?